kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,322 @@
+local constants = require "kong.constants"
+local clustering_utils = require "kong.clustering.utils"
+-- currently they are the same. But it's possible for we to drop support for old version of DP but keep support of CP
+local supported_services = require "kong.clustering.services.supported"
+local asked_services = require "kong.clustering.services.supported"
+local table_clear = require "table.clear"
+
+local time = ngx.time
+local var = ngx.var
+local log = ngx.log
+local ERR = ngx.ERR
+local NOTICE = ngx.NOTICE
+local _log_prefix = "[wrpc-clustering] "
+local table_concat = table.concat
+local lower = string.lower
+local pcall = pcall
+
+-- an optimization. Used when a not modified empty table is needed.
+local empty_table = {}
+
+local pairs = pairs
+local ipairs = ipairs
+local type = type
+local error = error
+
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+
+-- it's so annoying that protobuf does not support map to array
```

Comment:
this can be removed by not using a map, especially if you simplify the request (see comment in the .proto file)